### PR TITLE
Multi route support

### PR DESCRIPTION
**Pull request by Delawen**
Backend has now an experimental `v2` api that supports multiroutes. Not all endpoints are ported from `v1` yet.

Integration kind is not fully supported yet. This PR focused on Camel Routes, Kamelets, and Bindings.
Mixing different DSLs on the same file is not supported.

Superseeds https://github.com/KaotoIO/kaoto-backend/pull/633

### CODE REVIEW

**Comment by Delawen** [comment]
@lordrip :eyes:

File: api/pom.xml [hunk 1]
```diff
@@ -57,6 +57,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-smallrye-openapi</artifactId>
         </dependency>
+        <dependency>
```
Comment (Delawen) [inline api/pom.xml:60]: Added this because the discussion we had yesterday about caching things. So I thought `v2` of the API should experiment with caching on the backend side too.

Maybe I should separate in another PR? I just this dependency and an annotation.

File: api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+     * 🐱param dsl: String
+     * 🐱param integration: List<Integration>
+     *
+     * Idempotent operation that given an array of integrations, returns the corresponding CRDs.
+     */
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces("text/yaml")
+    @Path("/")
+    @CacheResult(cacheName = "api")
```
Comment (Delawen) [inline api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java:61]: Caching (see above comment)

File: api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    @Operation(summary = "Get Integration Object",
+            description = "Given the associated custom resource definition, returns the JSON object."
+                    + " This is an idempotent operation.")
+    public List<Integration> integration(
+            final @RequestBody String crd,
+            final @Parameter(description = "DSL to use. For example: "
+                    + "'Kamelet Binding'.")
+            @QueryParam("dsl") String dsl) {
+        List<Integration> integrations = new ArrayList<>();
```
Comment (Delawen) [inline api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java:92]: We should move all this to some service and keep the resources classes clean, but better not add more changes, it is already very long.

File: api/src/test/java/io/kaoto/backend/api/resource/v2/IntegrationsResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,393 @@
+package io.kaoto.backend.api.resource.v2;
```
Comment (Delawen) [inline api/src/test/java/io/kaoto/backend/api/resource/v2/IntegrationsResourceTest.java:1]: I copied the `v1` file here thinking at some point we will deprecate and remove `v1` and its tests.

At the end the round trip function.

File: api/src/test/java/io/kaoto/backend/api/resource/v2/IntegrationsResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+                .when()
+                .contentType("text/yaml")
+                .body(route)
+                .post("?dsl=" + parameters[0])
+                .then()
+                .statusCode(Response.Status.OK.getStatusCode());
+
+        assertEquals(res0.extract().body().asString(), res1.extract().body().asString());
+
+        var flows = Arrays.asList(res1.extract().body().as(Integration[].class));
```
Comment (Delawen) [inline api/src/test/java/io/kaoto/backend/api/resource/v2/IntegrationsResourceTest.java:377]: Note that now we return a list/array of Integrations.

File: api/src/test/resources/io/kaoto/backend/api/resource/amq-amq-multi.json [hunk 6]
```diff
@@ -0,0 +1,2798 @@
+[
```
Comment (Delawen) [inline api/src/test/resources/io/kaoto/backend/api/resource/amq-amq-multi.json:1]: JSON responses now are arrays/lists, not single objects. So I had to copy some resources files and added the `multi` string to distinguish.

File: api/src/test/resources/io/kaoto/backend/api/resource/kamelet-multi.yaml [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          uri: avro:https:hostname:999/message
+      - to:
+          uri: dropbox:put
+          parameters:
+            remotePath: '{{remotePath}}'
+            clientIdentifier: '{{clientIdentifier}}'
+            uploadMode: '{{uploadMode}}'
+            accessToken: '{{accessToken}}'
+
+---
```
Comment (Delawen) [inline api/src/test/resources/io/kaoto/backend/api/resource/kamelet-multi.yaml:79]: :smiling_imp: multi Kamelets on the same file MUAHAHAHA

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/camelroute/IntegrationDeploymentGeneratorService.java [hunk 8]
```diff
@@ -77,7 +78,20 @@ public String validationSchema() {
 
     @Override
     public String parse(final List<Step> steps, final Map<String, Object> metadata, final List<Parameter> parameters) {
-        return kdgs.getYAML(new Integration(steps, metadata, catalog), new IntegrationRepresenter());
+        return kdgs.getYAML(new Integration(
+                        steps != null ? steps : List.of(),
```
Comment (Delawen) [inline camel-route-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/camelroute/IntegrationDeploymentGeneratorService.java:82]: Noticed while testing that we were not very careful here. Strengthened it.

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/camelroute/IntegrationDeploymentGeneratorService.java [hunk 9]
```diff
@@ -107,6 +121,12 @@ public boolean appliesTo(final List<Step> steps) {
                         .anyMatch(Predicate.isEqual(s.getKind().toUpperCase())));
     }
 
+    @Override
+    public boolean appliesToFlows(List<StepParserService.ParseResult<Step>> flows) {
```
Comment (Delawen) [inline camel-route-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/camelroute/IntegrationDeploymentGeneratorService.java:125]: If one of the flows in the list of flows is from this DSL we **_assume_** that we can handle it.

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/CamelRouteStepParserService.java [hunk 10]
```diff
@@ -40,19 +40,29 @@ public String description() {
 
     @Override
     public ParseResult<Step> deepParse(final String input) {
+        // Right now we discard any flow that is not the first
```
Comment (Delawen) [inline camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/CamelRouteStepParserService.java:43]: Kept for `v1` until we deprecate.

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java [hunk 11]
```diff
@@ -81,6 +80,16 @@ public ParseResult<Step> deepParse(final String input) {
         return res;
     }
 
+    @Override
+    public List<ParseResult<Step>> getParsedFlows(String input) {
+        var res = new LinkedList<ParseResult<Step>>();
+        String[] splitCRDs = input.split(System.lineSeparator() + "---" + System.lineSeparator());
```
Comment (Delawen) [inline camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java:86]: Maybe this is too weak? Not sure how to make it better, to be honest. This could be in the middle of a long string text. Evil people will do evil things, you know.

File: kamelet-support/src/main/java/io/kaoto/backend/KamelPopulator.java [hunk 12]
```diff
@@ -118,7 +118,7 @@ public void populateKamelet(
                 def.setTitle(String.valueOf(map.getOrDefault("title", "")));
                 def.setDescription(String.valueOf(map.getOrDefault("description", "")));
                 def.setRequired((List<String>) map.getOrDefault("required", null));
-                def.setProperties((Map<String, KameletDefinitionProperty>) map.getOrDefault("required", null));
```
Comment (Delawen) [inline kamelet-support/src/main/java/io/kaoto/backend/KamelPopulator.java:121]: Caught this bug while adding more kamel examples. Oopsie.

File: kamelet-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/kamelet/KameletBindingDeploymentGeneratorService.java [hunk 13]
```diff
@@ -203,11 +209,32 @@ public Status getStatus(final CustomResource cr) {
         return s;
     }
 
+    @Override
+    public String parse(List<StepParserService.ParseResult<Step>> flows) {
+        StringBuilder sb = new StringBuilder();
+        flows.stream().forEachOrdered(stepParseResult -> {
+            if (!sb.isEmpty()) {
```
Comment (Delawen) [inline kamelet-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/kamelet/KameletBindingDeploymentGeneratorService.java:216]: Simple but effective.

File: kamelet-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/kamelet/KameletRepresenter.java [hunk 14]
```diff
@@ -211,6 +212,36 @@ public Node representData(final Object data) {
                                 DumperOptions.FlowStyle.BLOCK);
                     }
                 });
+
+        this.multiRepresenters.put(KameletDefinitionProperty.class,
+                new RepresentMap() {
```
Comment (Delawen) [inline kamelet-support/src/main/java/io/kaoto/backend/api/service/deployment/generator/kamelet/KameletRepresenter.java:217]: This was also missing, dumb us.

File: services-interfaces/src/main/java/io/kaoto/backend/api/service/deployment/generator/DeploymentGeneratorService.java [hunk 15]
```diff
@@ -55,8 +56,20 @@ public interface DeploymentGeneratorService {
      * If applies, the name will be the name used on the integration deployed.
      */
     @WithSpan
-    String parse(List<Step> steps, Map<String, Object> metadata,
-                 List<Parameter> parameters);
+    @Deprecated
```
Comment (Delawen) [inline services-interfaces/src/main/java/io/kaoto/backend/api/service/deployment/generator/DeploymentGeneratorService.java:59]: deprecated because we should discourage its use.

File: api/pom.xml (hunk 1 above)
Comment (Delawen) [reply api/pom.xml:60]: Ah great, github only shows the code above the comment. I always add comments on top of what I want people to see.

Comment (lordrip) [reply api/pom.xml:60]: I think that we could keep it in this very same pull request.

File: api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java (hunk 2 above)
Comment (lordrip) [reply api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java:61]: How would this cache behave? Is [this the correct "doc" for this annotation](https://es.quarkus.io/guides/cache)?

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java (hunk 11 above)
Comment (lordrip) [reply camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java:86]: yes, haters gonna hate, I know... Would something [like this](https://github.com/FasterXML/jackson-dataformats-text/issues/66#issuecomment-1386673293) help?

**Review by lordrip — APPROVED** [review]
LGTM 👍 

Just to confirm:

### 1. Response shape
given the following `YAML`
```yaml
- from:
    uri: null
    steps:
        - aggregate:
              aggregation-strategy: aggregate

- from:
    uri: null
    steps: []

- from:
    uri: null
    steps: []
```

we receive an array of routes, which is what I would expect to receive, but just wanted to confirm if this is correct or I should expect an object as you mentioned in another conversation?
image

### 2. YAML format
Sending 2 integrations in an array, we get the following routes:
image

would be possible / appropriate / useful to give an empty line between the routes? this is extremely 💅 (stylish)

**Comment by lordrip** [comment]
BTW, almost forgot, as soon as we get this merged, I'll switch the UI to use the new endpoints and interact only with the first route in the array. After that, I'll start incorporating individual functionality into the multi-route support.

**Comment by Delawen** [comment]
> BTW, almost forgot, as soon as we get this merged, I'll switch the UI to use the new endpoints and interact only with the first route in the array. After that, I'll start incorporating individual functionality into the multi-route support.

You like danger :smile_cat:

**Comment by Delawen** [comment]
Added an object wrapper to return instead of a plain array of flows. You were right.

File: api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java (hunk 2 above)
Comment (Delawen) [reply api/src/main/java/io/kaoto/backend/api/resource/v2/IntegrationsResource.java:63]: Yes, exactly that one.

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java (hunk 11 above)
Comment (Delawen) [reply camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java:86]: Let me try :)

**Comment by lordrip** [comment]
Alright, now it sends and receives an object like this:

```json5
{
  "flows": [ ... ],
  "properties: { ... }
}
```

**Review by lordrip — APPROVED** [review]

File: camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java (hunk 11 above)
Comment (Delawen) [reply camel-route-support/src/main/java/io/kaoto/backend/api/service/step/parser/camelroute/IntegrationStepParserService.java:86]: Not really working as expected.

**Comment by Delawen** [comment]
Adding an extra line between routes is breaking tests in an ugly way. I will leave that for a second PR.
